app/src/main/java/com/example/tasky/feature_authentication/presentation/login_screen/LoginViewModel.kt

```diff
@@ -83,13 +95,74 @@ class LoginViewModel @Inject constructor(
         }
     }
 
+    private fun validatePassword(password: String) {
+        val result = userDataValidator.validatePassword(password)
+        if(result.isValid) {
+            _state.update {
+                it.copy(
+                    isPasswordValid = true,
+                    passwordError = null
+                )
+            }
+        } else {
+            _state.update {
+                it.copy(
+                    isPasswordValid = false
+                )
+            }
+        }
+    }
+
     private fun signIn() {
+
         viewModelScope.launch {
-           val result = useCases.login(
-                _state.value.email,
-                _state.value.password
-            )
-            resultChannel.send(result)
+            val emailResult = userDataValidator.validateEmail(state.value.email)
+            val passwordResult = userDataValidator.validatePassword(state.value.password)
+
+            val hasError = listOf(
+                emailResult,
+                passwordResult
+            ).any{ !it.isValid }
+
+            if(hasError) {
+            _state.update {
+                it.copy(
+                    emailError = when(emailResult.emailError) {
+                        EmailError.EMAIL_EMPTY -> {
+                            emailErrorChannel.send(UiText.StringResource(R.string.EMAIL_EMPTY))
+                            "Email Empty"
+                        }
+                        EmailError.EMAIL_INVALID -> {
+                            emailErrorChannel.send(UiText.StringResource(R.string.EMAIL_INVALID))
+                            "Email Invalid"
+                        }
+                        null -> null
+                    },
+                    passwordError = when(passwordResult.passwordError) {
+                        PasswordError.PASSWORD_EMPTY -> {
+                            passwordErrorChannel.send(UiText.StringResource(R.string.PASSWORD_EMPTY))
+                            "Password Empty"
+                        }
+                        PasswordError.PASSWORD_INVALID -> {
+                            passwordErrorChannel.send(UiText.StringResource(R.string.PASSWORD_INVALID))
+                            "Password Invalid"
+                        }
+                        PasswordError.PASSWORD_TOO_SHORT -> {
+                            passwordErrorChannel.send(UiText.StringResource(R.string.PASSWORD_TOO_SHORT))
+                            "Password Too Short"
+                        }
+                        null -> null
+                    }
```

Comment:
What's the purpose of the dynamic string attached here? You already attach the string resource 🤔